tools/testsys-config/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+use bottlerocket_variant::Variant;
+pub use error::Error;
+use handlebars::Handlebars;
+use maplit::btreemap;
+use model::SecretName;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::collections::{BTreeMap, HashMap};
+use std::fs;
+use std::path::Path;
+pub type Result<T> = std::result::Result<T, error::Error>;
+
+/// Configuration needed to run tests
+#[derive(Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case")]
+pub struct TestConfig {
+    /// High level configuration for TestSys
+    pub test: Option<Test>,
+
+    #[serde(flatten, serialize_with = "toml::ser::tables_last")]
+    /// Configuration for testing variants
+    pub configs: HashMap<String, GenericConfig>,
+}
+
+impl TestConfig {
+    /// Deserializes a TestConfig from a given path
+    pub fn from_path<P>(path: P) -> Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        let path = path.as_ref();
+        let test_config_str = fs::read_to_string(path).context(error::FileSnafu { path })?;
+        toml::from_str(&test_config_str).context(error::InvalidTomlSnafu { path })
+    }
+
+    /// Deserializes a TestConfig from a given path, if it exists, otherwise builds a default
+    /// config
+    pub fn from_path_or_default<P>(path: P) -> Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        if path.as_ref().exists() {
+            Self::from_path(path)
+        } else {
+            Ok(Self::default())
+        }
+    }
+
+    /// Create a single config for the `variant` and `arch` from this config. `starting_config` is
+    /// used as the config with the highest precedence. If no starting config is available,
+    /// `Default::default()` can be used.
+    pub fn reduced_config<S>(
```

Comment:
`resolve_config`?

Can you better explain the purpose of the `starting_config` argument?

"Create a single config [...] from this config" is unclear.

Maybe it should be an `Option<GenericVariantConfig>`. It's not clear how to call the function if "no starting_config is available", but maybe what you're saying is that it is an optional argument.

Reply:
The technical term for the process is `reduce` so I'm not sure we want to change that. I have made `starting_config` an option.